### PR TITLE
structured code for app-attach

### PR DESCRIPTION
**Pull request by harshgondaliya**

### CODE REVIEW

**Comment by harshgondaliya** [comment]
@edwarnicke 
I have some questions and need your help for it.

1. Why are we using pointers in struct fields and not data variables?
Ex: type Attachment struct {
   ns *Namespace............}

2. I am passing context as parameter to all functions so that I can generate log messages. Is this fine?

3. Already the below two if conditions check whether the fd is for vppMqSegment or memfdSegment
if replyMsg.Msg.FdFlags&fdFlagVppMqSegment > 0 {
............
if replyMsg.Msg.FdFlags&fdFlagMemfdSegment > 0 {
Then why do we need to implement IsVppMqMemorySegment function?

4. For NewAttachment function, I am required to create MemorySegments (custom struct that we defined). MemorySegment struct has an attachment field. What value do I pass in that?

File: app-attach/namespace.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	}
+}
+
+func (ns *Namespace) Dial(ctx context.Context) (net.Conn, error){
+	socketAddr := fmt.Sprintf("/var/run/vpp/app_ns_sockets/%v", ns.id)
+   	udsConn, dErr := net.Dial("unixpacket", socketAddr)
+   	if dErr != nil {
+		log.Entry(ctx).Fatalln("Dial Error", dErr)
+   	}
+   	log.Entry(ctx).Infof("Connected to Unix Socket")
```
Comment (edwarnicke) [inline app-attach/namespace.go]: https://github.com/sirupsen/logrus

File: app-attach/namespace.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/edwarnicke/vpphelper"
+)
+var doOnce sync.Once
+type Connection interface {
+	api.Connection
+	api.ChannelProvider
+}
+
+type Namespace struct {
+    vppConn vpphelper.Connection
```
Comment (edwarnicke) [inline app-attach/namespace.go]: Golang supports [Duck Typing](https://medium.com/@matryer/golang-advent-calendar-day-one-duck-typing-a513aaed544d).

[vpphelper.Connection](https://github.com/edwarnicke/vpphelper/blob/3e6797de32c36a2eac2ac30e0932d04dc796fc6d/start.go#L36-L39) is just

```go
// Connection Combination of api.Connection and api.ChannelProvider
type Connection interface {
	api.Connection
	api.ChannelProvider
}
```

I would recommend you define your own 'Connection' similarly locally to your package.

The reason for this is that your eventual users may not be vpphelper users, and should not be forced to include vpphelper to use your library code.

File: app-attach/memory_segment.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+package main
+
+import (
+	"context"
+
+	"github.com/edwarnicke/log"
+	"github.com/justincormack/go-memfd"
+)
+type MemorySegment struct {
+    attachment *Attachment
```
Comment (edwarnicke) [inline app-attach/memory_segment.go]: Move the attachment member to the Worker type.

**Pull request merged**